### PR TITLE
replace os.path.getsize with  Path("file.txt").stat().st_size

### PR DESCRIPTION
**Pull request by doomedraven**
Hello is my first try to extend refurb. if i did something wrong let me know please. 
The ID is 154 bcz latest is 153 (according to `grep -r "code = " .| cut -d"=" -f 2|sort|uniq`)

### CODE REVIEW

**Review by dosisod — CHANGES_REQUESTED** [review]
Thank you @doomedraven for this PR! Everything looks good for the most part, minus the comments I left.

In addition, there are no tests for this code yet. I need to update the documentation for this, but in short, all you have to do is create a new file named `test/data/err_xxx.py` (where `xxx` is the error id), and add some code which will emit an error:

```python
from os.path import getsize

_ = getsize("filename")
```

Then, run `make update-tests`. This will create a file called `test/data/err_xxx.txt` with content similar to:

```
test/data/err_xxx.py:3:5 [FURBxxx]: Replace `os.path.getsize(x)` with `Path(x).stat().st_size`
```

Note that `make update-tests` only runs if the `.py` test file was changed, not if the check itself was updated. If you update the check but not the test file, you will have to run this instead:

```
refurb test/data/err_xxx.py --quiet > test/data/err_xxx.txt
```

I appreciate you putting in the effort to make a PR for Refurb, and for the most part, this is ready to be merged once the other notes are addressed!

(And sorry for the delay on this review, I have been on vacation for the past week :palm_tree:)

**Comment by doomedraven** [comment]
thank you for feedback. i got a question. Generated test shows `FURB146`, shouldn't it be `FURB155`?
`test/data/err_155.py:4:5 [FURB146]:`

File: refurb/checks/pathlib/getsize.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    case _:
+                        return
+
+                replace = f"Path(x).{new_name}"
+
+            errors.append(
+                ErrorInfo(
+                    node.line,
+                    node.column,
+                    f"Replace `os.path.getsize(x)` with `{replace}`",
```
Comment (dosisod) [inline refurb/checks/pathlib/getsize.py:75]: This is returning `os.path.getsize` for all function names, but should return the `os.path.whatever` to match the name in question.

File: refurb/checks/pathlib/getsize.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+}
+
+
+def check(node: CallExpr, errors: list[Error]) -> None:
+    match node:
+        case CallExpr(
+            callee=MemberExpr(fullname=fullname, name=name),
+            args=[arg],
+        ) if new_name := PATH_TO_PATHLIB_NAMES.get(fullname or ""):
```
Comment (dosisod) [inline refurb/checks/pathlib/getsize.py:52]: See my review comment above for context.

I would suggest changing this section to the following:

```python
PATH_TO_PATHLIB_NAMES = {
    "os.stat": "stat()",
    "os.path.getsize": "stat().st_size",
    "os.path.getatime": "stat().st_atime",
    "os.path.getmtime": "stat().st_mtime",
    "os.path.getctime": "stat().st_ctime",
}


def check(node: CallExpr, errors: list[Error]) -> None:
    match node:
        case CallExpr(
            callee=MemberExpr(fullname=fullname),
            args=[arg],
        ):
            normalized_name = fullname or ""

            for name in ("posix", "nt", "generic"):
                if normalized_name.startswith(name):
                    normalized_name = normalized_name.replace(name, "os.")

            new_name = PATH_TO_PATHLIB_NAMES.get(normalized_name)

            if not new_name:
                return
```

This basically will normalize the `posixpath`, `genericpath` and `ntpath` fields to `os.path`. This makes the lookup logic simpler and more self-explanatory.

File: test/data/err_155.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (dosisod) [inline test/data/err_155.py:1]: These are good test cases, but they don't cover all the cases (ie, passing a string/byte literal, etc). To get to 100% test coverage, add the following tests/lines:

```python
from pathlib import Path

os.path.getsize("filename")
os.path.getsize(b"filename")
os.path.getsize(Path("filename"))

# this should not match
os.path.getsize(1)
```

The last line checks that Refurb *doesn't* emit an error for that line. Reason being is that a lot of the `os.path` functions take a file descriptor (as an integer), whereas the `Path` function cannot, and thus we have to ignore this edge case.

Basically add these test cases, rerun `make update-tests`, and ensure the lines look correct (and that the last line with `getsize(1)` isn't in there).

**Review by dosisod — CHANGES_REQUESTED** [review]
Interesting, when I run it on my machine I get `155`. I can't see why it would be doing that, but if it is still doing that let me know and I will look into it further.

Also, the newly added `os.path.getatime` related functions are not showing up in the tests. When debugging, they are showing up as `genericpath.getatime`, so we probably need to expand these out like you did for the first one (ie, adding `genericpath`, `posixpath`, and `ntpath`). Doing that for each one might get tedious, so we might want to clean up/normalize the path name before we do the lookup.

File: refurb/checks/pathlib/getsize.py (hunk 2 above)
Comment (doomedraven) [reply refurb/checks/pathlib/getsize.py:52]: thank you

File: refurb/checks/pathlib/getsize.py (hunk 1 above)
Comment (doomedraven) [reply refurb/checks/pathlib/getsize.py]: done

**Comment by doomedraven** [comment]
about 146 was because I had added check size to my refurb to check my code. Thanks for feedback

**Comment by dosisod** [comment]
To fix the failing tests, add another line to the `err_155.py` file:

```python
os.path.getsize(__file__)
```

And run `make update-tests`. You don't have to add this for every `os.path` function, just one will do!

**Review by dosisod — APPROVED** [review]
Thank you for your contribution!

**Comment by doomedraven** [comment]
thanks for help to get this merged
